### ISSUE TITLE
New-AzRoleAssignment doesn't return error other than status code

### ISSUE THREAD

**Issue by abatishchev** [Azure/azure-powershell#19605]
### Description

1. Grant the permissions on an Azure subscription to a certain user, group, or service principal
2. Run a command to grant the permissions again

```powershell
New-AzRoleAssignment -ObjectId $id -RoleDefinitionName $role  -Scope "/subscriptions/$sub"
```

3. Observe an error:

> New-AzRoleAssignment: Operation returned an invalid status code 'Conflict'

3. Adding `-Verbose` doesn't change anything.
4. Only adding `-Debug` does!

### Issue script & Debug output

```PowerShell
DEBUG: ============================ HTTP REQUEST ============================

HTTP Method:
PUT

Absolute Uri:
https://management.azure.com/subscriptions/<sub id>/providers/Microsoft.Authorization/roleAssignments/<role id>?api-version=2020-08-01-preview

Headers:
x-ms-client-request-id        : e86e7fae-d39b-4f01-a34d-7034834acb4d
Accept-Language               : en-US

Body:
{
  "properties": {
    "roleDefinitionId": "/subscriptions/<sub id/providers/Microsoft.Authorization/roleDefinitions/<role id><role id>",
    "principalId": "<object id>",
    "principalType": "Group"
  }
}


DEBUG: ============================ HTTP RESPONSE ============================

Status Code:
Conflict

Headers:
Cache-Control                 : no-cache
Pragma                        : no-cache
x-ms-request-id               : 1e6c584a-9bc6-4450-92a5-b88e91b30091
X-Content-Type-Options        : nosniff
Strict-Transport-Security     : max-age=31536000; includeSubDomains
Set-Cookie                    : x-ms-gateway-slice=Production; path=/; secure; samesite=none; httponly
x-ms-ratelimit-remaining-subscription-writes: 1199
x-ms-correlation-request-id   : 54fa9c95-427f-4e62-8325-679b79241b0d
x-ms-routing-request-id       : WESTUS:20220923T191844Z:54fa9c95-427f-4e62-8325-679b79241b0d
Date                          : Fri, 23 Sep 2022 19:18:44 GMT

Body:
{
  "error": {
    "code": "RoleAssignmentExists",
    "message": "The role assignment already exists."
  }
}
```

### Environment data

```PowerShell
Name                           Value
----                           -----
PSVersion                      7.2.3
PSEdition                      Core
GitCommitId                    7.2.3
OS                             Microsoft Windows 10.0.19043
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Module versions

```PowerShell
ModuleType Version Name
---------- ------- ----
    Script 6.0.1   Az.Resources
```

### Error output

```PowerShell
HistoryId: 37

ErrorCategory     : CloseError: (:) [New-AzRoleAssignment], ErrorResponseException
ErrorDetail       :
InvocationInfo    : {New-AzRoleAssignment}
Line              : New-AzRoleAssignment -ObjectId "<object id>" -RoleDefinitionName "<role name>"  -Scope "/subscriptions/<sub id>"
Position          : At line:1 char:1
BoundParameters   : {}
UnboundParameters :
HistoryId         : 37

Resolve-AzError: Collection was modified; enumeration operation may not execute.
```

**Comment by dingmeng-xue** [comment]
It will help #19374.

**Comment by abatishchev** [comment]
Hi @dingmeng-xue, I don't see a solution over there, other than setting debug view preference. And I already do that by passing -Debug.

The volume of such issue manifests a design flaw and should be addressed. Turning on debug is not an option, it causes a too verbose output and isn't the standard practice to see **basic error message**. The current design is flawed and must be fixed, not worked around.

**Comment by dingmeng-xue** [comment]
@abatishchev , we need to change cmdlet implementation and show its full error message to end user. Current cmdlet swallows exception details.

**Comment by dingmeng-xue** [comment]
@Qarthak , thanks for your contribution. You can reference this page https://github.com/Azure/azure-powershell/blob/main/documentation/development-docs/azure-powershell-developer-guide.md

**Comment by abatishchev** [comment]
@dingmeng-xue should be the corresponding change be made in a centralized place so it would affect certain family of cmdlets?

**Comment by dingmeng-xue** [comment]
I'm not sure yet. Depends the implementation which varies across modules.

**Comment by Qarthak** [comment]
@dingmeng-xue I have built the environment on local without any errors. How should I proceed after this? Is there a particular file or directory that I should look for the bug in?

**Comment by dingmeng-xue** [comment]
@Qarthak , what file do you need? How to develop/debug Azure PowerShell cmdlet? or just source code around this bug?

**Comment by Qarthak** [comment]
Both would be helpful

**Comment by dingmeng-xue** [comment]
@Qarthak , please reference https://github.com/Azure/azure-powershell/blob/main/documentation/development-docs/azure-powershell-developer-guide.md. For this issue, I don't have insight yet. I believe it should come from below files or related files

- https://github.com/Azure/azure-powershell/blob/f296d277004fa0a99fa0c9dfa49696bec9763f2d/src/Resources/Resources/Models.Authorization/AuthorizationClient.cs
- https://github.com/Azure/azure-powershell/blob/main/src/Resources/Resources/RoleAssignments/NewAzureRoleAssignmentCommand.cs

**Comment by rachelxj-ms** [comment]
@dingmeng-xue this is centralized place to make the http request. 
https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/authorization/Microsoft.Azure.Management.Authorization/src/Generated/RoleAssignmentsOperationsExtensions.cs#L277
https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/authorization/Microsoft.Azure.Management.Authorization/src/Generated/RoleAssignmentsOperations.cs#L1058
![image](https://user-images.githubusercontent.com/66855967/200461928-9d1a3731-64ea-4a94-9374-633baddef53f.png)
The error message is in Exception body, but the Exception message doesn't have body appended, so it only prints "**New-AzRoleAssignment: Operation returned an invalid status code 'Conflict'**"

**Comment by rachelxj-ms** [comment]
https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/authorization/Microsoft.Azure.Management.Authorization/src/Generated/Models/ErrorResponseException.cs
There is property named "body" in this class, but it doesn't overwrite ToString() and print all details. 

I raised a ticket in azure-sdk-for-net, https://github.com/PowerShell/PowerShell/issues/18485

**Comment by rachelxj-ms** [comment]
@dingmeng-xue I got reply from azure-sdk-for-net:  The library you are using is our track 1 SDK and it is deprecated, please check our [deprecation announcement](https://azure.microsoft.com/en-us/updates/update-your-azure-sdk-libraries-of-net-and-golang-by-30-september-2023/).
As a replacement, you can use this package: https://www.nuget.org/packages/Azure.ResourceManager.Authorization

Not sure if it will fix this issue with new library yet.

**Comment by dingmeng-xue** [comment]
I believe current change doesn't require changes in SDK. We will work on it and you can reference it.

**Comment by dingmeng-xue** [comment]
It's long story. Upgrading to new version of SDK will make more potential issues and problems. It's not our direction now.

**Comment by rachelxj-ms** [comment]
Agreed. I have noticed the difficulty in upgrading to new SDK. It's much easier to fix case by case in each module, although it's not very decent.

**Comment by thancyya** [comment]
This is still an issue in version 6.6.0 of Az.Resources.

**Comment by MeirPeleg** [comment]
Still an issue... how long time MS needs to fix this?

**Comment by longb1** [comment]
still an issue 2025
